### PR TITLE
Time Selection Feature

### LINKED ISSUE #15123: Blogging Reminders v2 - time selection - Android
https://wpmobilep2.wordpress.com/2021/07/30/blogging-reminders-v2/

### PR DESCRIPTION
**Pull request by ravishanker**
Implemented time selection for blogging reminders

Fixes #15123 

To test:

- Go to site settings
- Find **Blogging reminders** setting and click on the item below it
- It should launch select days bottom sheet 
- bottom sheet should now have Notification time item, click on time
- It should show time selection dialog
- Select a time, AM, PM and time should reflect the selection
- De-select all days, time list item should disappear, then select a day, time should appear again

## Regression Notes
1. Potential unintended areas of impact
Reminder scheduling

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Unit testing

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by ravishanker** [comment]
Needs an update to FluxC in a separate PR [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2085)

**Review by renanferrari — COMMENTED** [review]
Thanks for the changes! I'm not sure if this is ready for review, but I've taken a quick look and it seems that the time picker is behaving a bit weirdly after I dismiss it by tapping on the outside:


https://user-images.githubusercontent.com/830056/128782688-30170e11-68cd-44dd-a578-87256b1fac40.mp4

Also, I noticed that the time formatting is not following my current locale?

Let me know if this is ready for a proper review, so I can take a deeper look at the code as well.

**Comment by ravishanker** [comment]
> Thanks for the changes! I'm not sure if this is ready for review, but I've taken a quick look and it seems that the time picker is behaving a bit weirdly after I dismiss it by tapping on the outside:
> video.7.mp4
> 
> Also, I noticed that the time formatting is not following my current locale?
> 
> Let me know if this is ready for a proper review, so I can take a deeper look at the code as well.

Thanks @renanferrari for a quick look.  Good catch on the issue.  Looks like if the time picker is cancelled, then it is not dismissing the dialog.  

Unfortunately, I couldn't use material MaterialTimePicker even after upgrading material version.  MaterialTimePicker is crashing due to bridge theme in use in the app.

I've now fixed the issue with Cancel of TimePicker.   You can may be review this after  I merge the [FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2085) PR first, then update this one.  Thanks

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt]: Instead of using the state to indicate if the dialog should be shown, I believe we could use a separate event, similar to how we do for the blogging reminder bottom sheet itself. You can take a look at the [`BloggingReminderUtils::observeBottomSheet`](https://github.com/wordpress-mobile/WordPress-Android/blob/a2034122225404ea3c567b144549ff1dd888dae6/WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderUtils.kt#L11) method to get an idea of how this would look like.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: Unless I'm missing something, I think we could use `setOnDismissListener` here instead, as that would probably cover all the cases we need.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: It seems that we're not using these arguments anywhere yet. I imagine you're planning to use them to pass the currently selected time to the dialog? If so, I think that's the way to go 👍

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt]: It looks like `isInvisible` is not being used anywhere?

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersUiModel.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersUiModel.kt:13]: We should probably use `LocalTime` here instead – it can be constructed pretty easily with `LocalTime.of(hour, minute)` and it would make formatting easier (left a [comment below](https://github.com/wordpress-mobile/WordPress-Android/pull/15128#discussion_r686254015) with an example).

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersUiModel.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersUiModel.kt]: If we use `LocalTime`, we could then have something like:

```kotlin
return LocalTime.of(hour, minute).format(DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT))
```

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt]: We should probably set the click listener on the whole view, instead of just the `TextView`. Otherwise, the touch target gets really small and easy to miss.

**Review by renanferrari — COMMENTED** [review]
Thanks for the changes! After testing, it seems that the dialog behavior is still there if we press the back button. I've taken a look at the code and left a few suggestions below. Let me know what you think.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt]: My first implementation was using a separate event, then changed it to the current one, assuming we would show timepicker in the next screen like iOS and not on top of day selection.  Given it is viewholders and recylerview, it is playing up with the events.  I'll revert to the separate event and see if it works better

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: Tried that one and also other listeners,  The OK and Cancel button on the dialog seem to fire the same onTimeset / onTimeChange callbacks.  MaterialTimePicker has all these distinct listeners, unfortunately can't use it.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: Yes, once I merge the FluxC PR

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt]: Need to seek clarification on this one.  We had discussion if you remember, whether the time selection should always be displayed or only after a day selection is made.  If we decide to default to display always, then this can be removed

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt]: I was wondering the same, should the padding needs to added here around time or just make the whole thing touch target.  I'll change to the whole view for now.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt]: Update it now, to be visible only when at least a day is selected

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt]: Update touch target to the list item

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: Removed the arguments now, no longer needed as time and hour are available after FluxC update

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt]: Update to implement as per your suggestion.  Thanks

File: build.gradle
Comment (ravishanker) [inline build.gradle:136]: NOTE to self:  Update this to FluxC version after merge

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (renanferrari) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: >The OK and Cancel button on the dialog seem to fire the same onTimeset / onTimeChange callbacks

Hmm 🤔 I actually couldn't reproduce that. I commented out all the lines above and the picker seems to only fire the `OnTimeSetListener` when I click OK. Maybe I'm missing something?

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (renanferrari) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: I believe we should still pass the currently selected time to the dialog. If we don't, the time picker will display a different time from the one that's being displayed on the bottom sheet:


https://user-images.githubusercontent.com/830056/129103685-a9fbb2cd-ab2f-4e20-b06a-1fcd0fea8b4d.mp4

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt
Comment (renanferrari) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt]: I'm not sure what's the best approach here, but the way we are currently hiding the item looks quite weird IMHO:

https://user-images.githubusercontent.com/830056/129104115-b58e87e0-2520-4219-937a-28fd23ed32b1.mp4

If we want to hide that, we should consider conditionally adding this [`TimeItem`](https://github.com/wordpress-mobile/WordPress-Android/blob/2a90f3d7507fbcd2c95dc4ac9aa7a0009e59b17d/WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt#L52-L56) to the list instead of hiding its content. Personally, I'd probably just default to always show it, though.

Either way, we should probably ask for a design review here. You can do that by adding the "Needs design review" label and/or pinging a designer yourself.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt
Comment (renanferrari) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt]: Making the whole view the target seems the right thing to do 👍 I'd just make sure to add the selectable background to the view, so we can see ripple animation when we touch it.

**Review by renanferrari — COMMENTED** [review]
Left a few more comments above. Let me know if you need a hand with anything!

File: WordPress/src/main/java/org/wordpress/android/workers/reminder/ReminderScheduler.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/workers/reminder/ReminderScheduler.kt]: Instead of adding new mutable variables here, we should probably either add a `time` argument to the `schedule` function or refactor `ReminderConfig` to handle `LocalDateTime`s instead of `LocalDate`s. The first option seems simpler.

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt]: We shouldn't need to ever set this hardcoded time 🤔

**Review by renanferrari — COMMENTED** [review]
Sorry, just noticed a few more things!

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: I mean cancel wasn't working properly before I changed to utils implementation.  Cleaned up everything now

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderTimePicker.kt]: Updated to reflect current time selection on the picker after merging FluxC PR

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersItem.kt]: Yeah, Chris is AFK, will ask next week when he returns.  For now, made TimeItem conditional.  Thanks

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewHolder.kt]: added selectableItemBackground.  Thanks

File: WordPress/src/main/java/org/wordpress/android/workers/reminder/ReminderScheduler.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/workers/reminder/ReminderScheduler.kt]: Updated schedule function with time arguments after merging FluxC PR.  Thanks

File: WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt
Comment (ravishanker) [reply WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/DaySelectionBuilder.kt]: Cleaned up, after merging FluxC PR.   Thanks

**Review by renanferrari — APPROVED** [review]
Great work!

LGTM :shipit: 

Feel free to merge this after resolving the merge conflicts. I think we can address any design issues separately as well, but it's up to you.

**Comment by osullivanchris** [comment]
Hey @renanferrari @ravishanker  I had a think about the time display when the user has not selected any days yet. 

What if we display it but disabled until days are selected? Then we follow the same rational that its not interactive before days are selected, but, the UI doesn't jump when it becomes active. WDYT?

The only thing I wasn't sure is if we should display the default time, or no time, before a day is selected.

notification time

I also thought Ravi had a nice idea about using a button style for the time to make it more clear that its interactive. Here's a quick mock. I think it looks pretty nice but I'm not sure if its ok to put a text button in a list item like this? 

alt

**Comment by osullivanchris** [comment]
As discussed elsewhere, we have good reason for 'hiding' rather than disabling the time initially. So we can ignore my above suggestion for now!

**Pull request merged**
